### PR TITLE
[Serializer] Add datetimezone normalizer

### PR DESCRIPTION
**Pull request by jewome62**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30145
| License       | MIT
| Doc PR        | not yet

### CODE REVIEW

**Review by maxhelias — APPROVED** [review]

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 -----
 
  * added the list of constraint violations' parameters in `ConstraintViolationListNormalizer`
+ * added support for serializing objects that implement `DateTimeZone`
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: added support for serializing `DateTimeZone` objects

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+
+/**
+ * Normalizes an object {@see \DateTimeZone} to a timezone string.
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: `Normalizes a {@see \DateTimeZone} object to a timezone string.`

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+     *
+     * @throws InvalidArgumentException
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        if (!$object instanceof \DateTimeZone) {
+            throw new InvalidArgumentException('The object must implement the "\DateTimeZone".');
+        }
+
+        return $object->getName();
```
Comment (maidmaid) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: Shoud we also return the offset ?

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+{
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        if (!$object instanceof \DateTimeZone) {
+            throw new InvalidArgumentException('The object must implement the "\DateTimeZone".');
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: The object must be an instance of "\DateTimeZone".

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws NotNormalizableValueException
+     */
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        if ('' === $data || null === $data) {
+            throw new NotNormalizableValueException('The data is either an empty string or null, you should pass a string that can be parsed with the passed format or a valid DateTimeZone string.');
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: I do not understand the "can be parsed with the passed format" part here.

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php (hunk 4 above)
Comment (maidmaid) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: (Why this exception if `supportsNormalization()` guarantees that `$object` is an instance of DateTimeZone ?)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php (hunk 3 above)
Comment (jewome62) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: the DateTimeZone is more precise and allows to find the data. the offset just returns the time difference for a DateTime (first argument)

File: src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php (hunk 4 above)
Comment (jewome62) [reply src/Symfony/Component/Serializer/Normalizer/DateTimeZoneNormalizer.php]: I have reproduce the same test from DateTimeNormalizer and DateIntervalNormalizer
I think that's implemented if normalize is call without call support before

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jewome62.

**Pull request merged**